src/Language/Rust.ts

```diff
@@ -0,0 +1,312 @@
+    }
+
+    protected topLevelNameStyle(rawName: string): string {
+        return camelCaseStyle(rawName);
+    }
+
+    private nullableRustType = (t: Type, withIssues: boolean): Sourcelike => {
+        return ["Option<", this.breakCycle(t, withIssues), ">"];
+    };
+
+    protected isImplicitCycleBreaker(t: Type): boolean {
+        const kind = t.kind;
+        return kind === "array" || kind === "map";
+    }
+
+    private rustType = (t: Type, withIssues: boolean = false): Sourcelike => {
+        return matchType<Sourcelike>(
+            t,
+            _anyType => maybeAnnotated(withIssues, anyTypeIssueAnnotation, "serde_json::Value"),
+            _nullType => maybeAnnotated(withIssues, nullTypeIssueAnnotation, "Option<serde_json::Value>"),
+            _boolType => "bool",
+            _integerType => "i64",
+            _doubleType => "f64",
+            _stringType => "String",
+            arrayType => ["Vec<", this.rustType(arrayType.items, withIssues), ">"],
+            classType => this.nameForNamedType(classType),
+            mapType => ["HashMap<String, ", this.rustType(mapType.values, withIssues), ">"],
+            enumType => this.nameForNamedType(enumType),
+            unionType => {
+                const nullable = nullableFromUnion(unionType);
+
+                if (nullable) return this.nullableRustType(nullable, withIssues);
+
+                const [hasNull,] = removeNullFromUnion(unionType);
+
+                const isCycleBreaker = this.isCycleBreakerType(unionType);
+
+                const name = isCycleBreaker ? ["Box<", this.nameForNamedType(unionType), ">"] : this.nameForNamedType(unionType);
+
+                return hasNull
+                    ? ["Option<", name, ">"] as Sourcelike
+                    : name;
+            }
+        );
+    };
+
+    private breakCycle = (t: Type, withIssues: boolean): any => {
+        const rustType = this.rustType(t, withIssues);
+        const isCycleBreaker = this.isCycleBreakerType(t);
+
+        return isCycleBreaker ? ["Box<", rustType, ">"] : rustType;
+    }
+
+    private emitStructDefinition = (c: ClassType, className: Name): void => {
+        this.emitLine("#[derive(Serialize, Deserialize)]");
+
+        const structBody = () => this.forEachClassProperty(c, "none", (name, jsonName, prop) => {
+
+            const escapedName = rustStringEscape(jsonName);
+
+            this.emitLine("#[serde(rename = \"", escapedName, "\")]");
+
+            this.emitLine(name, ": ", this.breakCycle(prop.type, true), ",");
+        });
+
+        this.emitBlock(["pub struct ", className], structBody);
+    };
+
+    private emitBlock = (line: Sourcelike, f: () => void): void => {
+        this.emitLine(line, " {");
+        this.indent(f);
+        this.emitLine("}");
+    };
+
+    private emitUnion = (u: UnionType, unionName: Name): void => {
+        const isMaybeWithSingleType = nullableFromUnion(u);
+
+        if (isMaybeWithSingleType) {

```

Comment:
Please never use `null` or `undefined` as a boolean - compare explicitly, i.e. `x === null`.  JavaScript will do stupid things like also treat empty strings as `false`, so you think you're checking whether it's `null`, but you're really checking whether it's `null`, or an empty string, or whatever else JS stupidly thinks are falsey.